File: code/day7.R
library(purrr)
library(stringr)

parse_rules <- function(file) {
  lines <- readLines(file)
  rulebook <- list()
  for (line in lines) {
    rule <- unlist(strsplit(line, " contain "))
    rulebook[[str_extract(rule[1], ".+(?= bags)")]] <- bag_amount(rule[2])
  }
  rulebook
}

bag_amount <- function(line) {
  sp <- unlist(strsplit(line, ", "))
  count <- as.numeric(str_extract(sp, "\\d"))
  names(count) <- trimws(str_extract(sp, "(?<=\\d ).+(?= bag|bags)"))
  count
}

shiny_gold <- function(bag) {
  if ("shiny gold" %in% bag) {
    return(TRUE)
  } else {
    return(any(map_lgl(names(rules[[bag]]), shiny_gold)))
  }
}

count_bags <- function(bag) {
  if (is.na(rules[[bag]][1])) {
    return(0)
  } else {
    return(sum(map2_dbl(names(rules[[bag]]), rules[[bag]], ~ .y + .y * count_bags(.x))))
  }
}

rules <- parse_rules("data/day7")

# puzzle 1
sum(map_lgl(names(rules)[names(rules) != "shiny gold"], shiny_gold))

# puzzle 2
count_bags("shiny gold")

File: code/day7.jl
function read_rules(file)
	lines = readlines(file)
	rules = Dict()
	for line in lines
		bag, contents = split(line, " bags contain ")
		if occursin("no other", contents)
			rules[bag] = nothing
			continue
		end
		rules[bag] = Dict{String, Int}()
		for value in split(contents, ", ")
			n, style = match(r"(\d+\s)([a-z]+ [a-z]+)", value).captures
			rules[bag][style] = parse(Int, n)
		end
	end
	rules
end

function contains_gold(rules, bag)
	 if rules[bag] === nothing
		 return false
	 end
	 for contents in collect(keys(rules[bag]))
		 if (contents == "shiny gold") || containsGold(rules, contents)
			 return true
		 end
	 end
	 return false
 end

 function count_bags(rules, color)
     contents = rules[color]
     if contents === nothing
         return 0
     else
         return sum([
             contents[key] * solve(rules, key) + contents[key]
             for key in collect(keys(contents))
         ])
     end
 end

rules = read_rules("data/day7")

# puzzle 1
[contains_gold(rules, bag) for bag in collect(keys(rules))] |> sum

# puzzle 2
count_bags(rules, "shiny gold")
